semantic/src/main/kotlin/tools/samt/semantic/Package.kt

```diff
@@ -0,0 +1,73 @@
+package tools.samt.semantic
+
+import tools.samt.parser.BundleIdentifierNode
+import tools.samt.parser.IdentifierNode
+import tools.samt.parser.NamedDeclarationNode
+
+class Package(val name: String) {
+    val subPackages: MutableList<Package> = mutableListOf()
+
+    val records: MutableList<RecordType> = mutableListOf()
+    val enums: MutableList<EnumType> = mutableListOf()
+    val services: MutableList<ServiceType> = mutableListOf()
+    val providers: MutableList<ProviderType> = mutableListOf()
+    val consumers: MutableList<ConsumerType> = mutableListOf()
+    val aliases: MutableMap<String, TypeReference> = mutableMapOf()
+
+    val types: MutableMap<String, Type> = mutableMapOf()
+
+    fun resolveType(name: BundleIdentifierNode): Type? {
+        val parts = name.components.map { it.name }
+        return resolveType(parts)
+    }
+
+    private fun resolveType(components: List<String>, offset: Int = 0): Type? {
```

Comment:
You could remove the offset parameter and just truncate the list

Reply:
But that would create many new list instances which is why I chose to use an offset. I don't think the code would be significantly more readable given just a components list and a `drop(1)` instead

Reply:
Lists are immutable and thus probably have some sort of COW optimization. Is there something akin to a Slice in Kotlin?

Reply:
There is a way to have a COW optimized version, but you still create some wrapper instances. I pasted the adjusted code below, given we should optimize on readability do you think this is more readable?

```kotlin
private fun resolveType(components: List<String>): Type? {
    val subPackage = subPackages.find { it.name == components.first() }

    if (components.size == 1) {
        return subPackage?.let { PackageType(it) } ?: types[components.first()]
    }
    return subPackage?.resolveType(components.subList(1, components.size))
}
```

Reply:
Not sure about that middle thing there but I'd prefer this version.